lib/internal/modules/esm/loader.js

```diff
@@ -168,100 +212,250 @@ class Loader {
     const job = new ModuleJob(this, url, evalInstance, false, false);
     this.moduleMap.set(url, job);
     const { module } = await job.run();
+
     return {
       namespace: module.getNamespace(),
     };
   }
 
-  async import(specifier, parent) {
-    const job = await this.getModuleJob(specifier, parent);
-    const { module } = await job.run();
-    return module.getNamespace();
+  async getModuleJob(specifier, parentURL) {
+    const { format, url } = await this.resolve(specifier, parentURL);
```

Comment:
🤔 I think the return from `this.resolve()` (aka `esmLoader.resolve()`) is safe from that? Were you maybe thinking `this.resolve` is a reference to the user's custom `resolve` hook? If so, we'd need to check own props of that/those return values (within `this.resolve()`); but I think that's needless effort: a user couldn't do this by accident, and doing it wouldn't break—just lead to potentially unexpected behaviour.

Reply:
My opinion would be to leave it: the more smoke, the easier to notice a fire.

I would MUCH rather document a general warning that node is vulnerable to prototype pollution from userland than document each instance a vulnerability is present. I'm quite confident the vast majority of users have no idea userland is not sandboxed and can affect node this way.

Reply:
If the hook returns an object like this `{ url: 'file:///dev/null' }`, `format` is nullish and everything may looks like it all works. But then if some user code mutates the object prototype (e.g.: `Object.prototype.format = function format() { return JSON.stringify(this); }`), or if new methods are added by TC39 in the future, it makes the code unreliable.

Reply:
The prototype pollution possibility is user folly (so see above).

> if new methods are added by TC39 in the future

Ahh, good point. This I would be concerned about and think warrants an explicit check—Node would break through no fault of the user (so something we should handle).

Reply:
> We might want to restrict format and url to be own properties of the resolver to avoid prototype pollution.

After a bit more of thinking, I'm not sure that's the right solution. I can see a hook author may want to use custom classes as return value, and which would use prototype inheritance to define `format` and `url`, so forcing to use own properties would be detrimental to them.

I don't know if I have the right answer here, after a bit of digging, I can see there are already traps like this in the ECMAScript spec (e.g.: if you set `Object.prototype.writable = false`, it will impact all `Object.defineProperty` calls that do not define `writable` explicitly). I see two ways of handling that:
- Document that it is recommended to always define both properties explicitly (or to use null prototype objects): there's no TC39 proposal to add any method using any of those names, it's seems like low risk tbh, so I think that would be acceptable.
- Instead of accepting nullish values, only accept `null` or a string: that would force hook authors to explicitly define the properties somewhere in the prototype chain instead of relying on `Object.prototype.format === undefined`.

In any case, this can be settled after this PR, but probably a discussion worth having before calling the feature stable.

Reply:
Some context from the meeting where optionally supporting `format` in `resolve()` was decided: the agreement to add it to `resolve()` in addition to `load()` (where it's required) was predicated on a user being able to omit it. I think that means options 1 & 2 above are precluded without revisiting.

The `in` operator sounds like a good happy medium (but as you mentioned, doesn't protect against future addition of those properties to `Object.prototype`). If we're concerned with that potential, that sounds like an ideal case for abstraction—a utility function like `hasCustomProperty(obj, propName)`. Until it's a problem, do a simple check; if ever we come to that bridge, somehow confirm the property is not coming from `Object.prototype`.